File: app/src/main/java/uk/co/sundroid/activity/data/fragments/dialogs/date/DatePickerFragment.java
package uk.co.sundroid.activity.data.fragments.dialogs.date;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.Fragment;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.DatePicker;
import android.widget.DatePicker.OnDateChangedListener;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import uk.co.sundroid.R.id;
import uk.co.sundroid.R.layout;

import java.util.Calendar;
import java.util.TimeZone;

import static java.util.Calendar.*;

public class DatePickerFragment extends DialogFragment implements DialogInterface.OnClickListener, OnDateChangedListener, OnSeekBarChangeListener {

    private Calendar calendar;
    private DatePicker datePicker;
    private SeekBar dateSeeker;

    @FunctionalInterface
    public interface OnDateSelectedListener {
        void onDateSet(int year, int month, int date);
    }

    public static DatePickerFragment newInstance(Calendar calendar) {
        DatePickerFragment fragment = new DatePickerFragment();
        Bundle args = new Bundle();
        args.putIntArray("ymd", new int[] { calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH) });
        args.putString("tz", calendar.getTimeZone().getID());
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        int[] ymd = getArguments().getIntArray("ymd");
        String tz = getArguments().getString("tz");
        calendar = Calendar.getInstance();
        if (ymd != null && tz != null) {
            calendar.setTimeZone(TimeZone.getTimeZone(tz));
            calendar.set(ymd[0], ymd[1], ymd[2]);
        }
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        setRetainInstance(true); // TODO May not be required
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        View view = View.inflate(getActivity(), layout.dialog_datepicker, null);
        datePicker = view.findViewById(id.datePicker);
        datePicker.init(calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH), this);
        builder.setView(view);
        builder.setTitle("Set date");

        dateSeeker = view.findViewById(id.dateSeeker);
        dateSeeker.setOnSeekBarChangeListener(this);
        dateSeeker.setMax(365);
        dateSeeker.setProgress(calendar.get(DAY_OF_YEAR));

        builder.setPositiveButton("Set", this);
        builder.setNeutralButton("Today", this);
        builder.setNegativeButton("Cancel", this);
        return builder.create();
    }

    @Override
    public void onDateChanged(DatePicker datePicker, int year, int month, int day) {
        if (dateSeeker != null && calendar != null) {
            Calendar cal = Calendar.getInstance(calendar.getTimeZone());
            cal.set(datePicker.getYear(), datePicker.getMonth(), datePicker.getDayOfMonth());
            dateSeeker.setProgress(cal.get(DAY_OF_YEAR));
        }
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int dayOfYear, boolean fromUser) {
        if (fromUser && datePicker != null && calendar != null) {
            dayOfYear = Math.max(1, dayOfYear);
            Calendar cal = Calendar.getInstance(calendar.getTimeZone());
            cal.set(Calendar.DAY_OF_YEAR, dayOfYear);
            datePicker.updateDate(datePicker.getYear(), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH));
        }
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int button) {
        Fragment target = getTargetFragment();
        if (target != null && target instanceof OnDateSelectedListener) {
            if (button == DialogInterface.BUTTON_NEUTRAL) {
                Calendar today = Calendar.getInstance();
                ((OnDateSelectedListener)target).onDateSet(today.get(YEAR), today.get(MONTH), today.get(DAY_OF_MONTH));
            } else if (button == DialogInterface.BUTTON_POSITIVE) {
                ((OnDateSelectedListener)target).onDateSet(datePicker.getYear(), datePicker.getMonth(), datePicker.getDayOfMonth());
            }
        }
        dismiss();
    }

}

File: app/src/main/kotlin/uk/co/sundroid/activity/data/fragments/dialogs/date/DatePickerFragment.kt
package uk.co.sundroid.activity.data.fragments.dialogs.date

import android.app.AlertDialog
import android.app.Dialog
import android.app.DialogFragment
import android.os.Bundle
import android.view.View
import android.widget.DatePicker
import android.widget.DatePicker.OnDateChangedListener
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import uk.co.sundroid.R
import uk.co.sundroid.R.layout
import java.util.*
import java.util.Calendar.*

class DatePickerFragment : DialogFragment(), OnDateChangedListener, OnSeekBarChangeListener {

    private var calendar: Calendar = Calendar.getInstance()
    private var datePicker: DatePicker? = null
    private var dateSeeker: SeekBar? = null

    @FunctionalInterface
    interface OnDateSelectedListener {
        fun onDateSet(year: Int, month: Int, date: Int)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        restore(arguments)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        Companion.save(calendar, outState)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        restore(savedInstanceState)

        val view = View.inflate(activity, layout.dialog_datepicker, null)

        datePicker = view.findViewById(R.id.datePicker)
        datePicker?.apply {
            init(calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH), this@DatePickerFragment)
        }
        dateSeeker = view.findViewById(R.id.dateSeeker)
        dateSeeker?.apply {
            setOnSeekBarChangeListener(this@DatePickerFragment)
            max = 365
            progress = calendar.get(DAY_OF_YEAR)
        }

        return AlertDialog.Builder(activity).apply {
            setView(view)
            setTitle("Set date")
            setPositiveButton("Set", { _, _ -> set(calendar) })
            setNeutralButton("Today", { _, _ -> set(Calendar.getInstance()) })
            setNegativeButton("Cancel", null)
        }.create()
    }

    override fun onDateChanged(datePicker: DatePicker, year: Int, month: Int, day: Int) {
        calendar.set(datePicker.year, datePicker.month, datePicker.dayOfMonth)
        dateSeeker?.progress = calendar.get(DAY_OF_YEAR)
    }

    override fun onProgressChanged(dateSeeker: SeekBar, dayOfYear: Int, fromUser: Boolean) {
        if (fromUser) {
            calendar.set(Calendar.DAY_OF_YEAR, Math.max(1, dayOfYear))
            datePicker?.updateDate(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH))
        }
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {}

    override fun onStopTrackingTouch(seekBar: SeekBar) {}

    private fun set(calendar: Calendar) {
        val target = targetFragment
        if (target is OnDateSelectedListener) {
            (target as OnDateSelectedListener).onDateSet(calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH))
        }
    }

    private fun restore(bundle: Bundle?) {
        if (bundle != null) {
            val ymd = bundle.getIntArray("ymd")
            val tz = bundle.getString("tz")
            if (ymd != null && tz != null) {
                calendar.timeZone = TimeZone.getTimeZone(tz)
                calendar.set(ymd[0], ymd[1], ymd[2])
            }
        }
    }

    companion object {
        fun newInstance(calendar: Calendar): DatePickerFragment {
            return DatePickerFragment().apply {
                arguments = save(calendar)
            }
        }

        private fun save(calendar: Calendar, bundle: Bundle? = Bundle()): Bundle? {
            return bundle?.apply {
                putIntArray("ymd", intArrayOf(calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH)))
                putString("tz", calendar.timeZone.id)
            }
        }
    }

}